mlx_lm/server.py

```diff
@@ -145,6 +147,146 @@ def process_message_content(messages):
             message["content"] = ""
 
 
+class LRUPromptCache:
+
+    @dataclass
+    class CacheEntry:
+        prompt_cache: List[Any]
+        count: int
+
+    @dataclass
+    class SearchResult:
+        model: Any
+        exact: List[int]
+        shorter: List[int]
+        longer: List[int]
+        common_prefix: int
+
+    def __init__(self, max_size: int = 10):
+        self.max_size = max_size
+        self._cache = {}
+        self._lru = deque()
+
+    def _search(self, model, tokens):
+        """Search the cache for a prompt cache. Return exact or close match."""
+        if model not in self._cache:
+            return self.SearchResult(model, None, None, None, 0)
+
+        current = self._cache[model]
+        last_cache_index = -1
+        index = 0
+
+        while index < len(tokens) and tokens[index] in current:
+            current = current[tokens[index]]
+            if "cache" in current:
+                last_cache_index = index
+            index += 1
+
+        # Exact match no need to search for longer or shorter caches
+        if last_cache_index == len(tokens) - 1:
+            return self.SearchResult(model, tokens, None, None, 0)
+
+        # Find the shorter cache
+        shorter = None
+        if last_cache_index > 0:
+            shorter = tokens[: last_cache_index + 1]
+
+        # Check for caches that are longer
+        longer = None
+        common_prefix = index
```

Comment:
~What's the purpose to check for longer if you found a shorter? Is the idea to use a different policy in the future?~ nvm I see the condition now.